content/2025-09-01-gsoc-ubsan-trap-messages.md

```diff
@@ -0,0 +1,123 @@
+
+TODO: Show example of normal userspace UBSan and how it outputs the problem
+TODO: show example of trapping UBSan. Both with and without the debugger. This will let you illustrate the problem of it being hard to understand what happened, even in the debugger. This gives you the motivation for your work which is currently missing.
+
+## Human readable descriptions of UBSan traps in LLDB
+
+During my GSoC project I implemented support for displaying human readable descriptions of UBSan traps in LLDB to improve the debugging experience.
+
+The approach used is based on how `__builtin_verbose_trap` is implemented.
+
+This was done by inserting a fake frame into debug-info, which is formatted like so: `__clang_trap_msg$<Category>$<TrapMessage>`. This specific format
+is recognized and can be encoded by LLDB, which resembles what is used for `__builtin_verbose_trap`. 
+
+In both cases, the compiler encodes the trap's context by emitting the trap instruction inside an artificial function with a specially-formatted name (e.g., `__clang_trap_msg$<Category>$<Message>`). When the trap occurs, LLDB's Verbose Trap StackFrame Recognizer identifies this special name in the debug info and uses the extracted category and message to generate the user-friendly stop reason. By adopting this existing protocol, this feature ensures robust and consistent behavior within the debugger.
+
+Take for example this erroneous program, `foo.c`:
+
+```
+#include <limits.h>
+
+int main() { return INT_MAX + 1; }
+```
+
+When run with these flags before my change:
+`$ clang -fsanitize=signed-integer-overflow -fsanitize-trap=signed-integer-overflow foo.c`
+
+Debug info would not provide a stopping reason, leaving the user possibly confused. 
+
+After the changes, this line is emitted in LLVM IR:
+`!18 = distinct !DISubprogram(name: "__clang_trap_msg$Undefined Behavior Sanitizer$signed integer addition overflow in '2147483647 + 1'", scope: !1, file: !1, type: !19, flags: DIFlagArtificial, spFlags: DISPFlagDefinition, unit: !0)`
+
+Which looks something like this in LLDB: 
+
+`stop reason = Undefined Behavior Sanitizer: signed integer addition overflow in '2147483647 + 1'`
+
+Previously, the program would stop execution, but the user would not know why. With the new feature, the stop reason is apparent. 
+
+The `-fsanitize-debug-trap-reasons` flag [1] enables trap messages for UBSan, which provides context for stop reasons in trapping UBSan.
```

Comment:
I would remove this and say something along the lines of: "one concern was debug-info size. Here are the results of my bloaty investigations. To provide an escape hatch, we made this an on-by-default flag called `-fsanitize-debug-trap-reasons`